### PR TITLE
Let kubelet support lxcfs

### PR DESCRIPTION
**Pull request by fanux**
**What type of PR is this?**
> /kind feature

**What this PR does / why we need it**:
Support lxcfs, a better kubelet for java apps.

**Special notes for your reviewer**:
/sig kubelet

### CODE REVIEW

**Comment by fanux** [comment]
# kubelet with lxcfs
A better kubelet for java applications.

# install lxcfs 
```
wget https://copr-be.cloud.fedoraproject.org/results/ganto/lxd/epel-7-x86_64/00486278-lxcfs/lxcfs-2.0.5-3.el7.centos.x86_64.rpm
yum install lxcfs-2.0.5-3.el7.centos.x86_64.rpm 
systemctl enable lxcfs
systemctl start lxcfs
```

# Test 
```
mkdir /root/static
touch  /root/static/pod.yaml
```

```
[root@izj6c996fvahbz1keqg2iqz ~]# cat static/pod.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: myapp-pod
  labels:
    app: myapp
spec:
  containers:
  - name: myapp-container
    image: ubuntu:18.04
    command: ['sh', '-c', 'echo Hello Kubernetes! && sleep 3600']
    resources:
      requests:
        memory: "1Gi"
        cpu: "500m"
      limits:
        memory: "1Gi"
        cpu: "500m"
```

```
./kubelet --pod-manifest-path /root/static --cgroup-driver systemd  --runtime-cgroups=/systemd/system.slice --kubelet-cgroups=/systemd/system.slice
```

```
docker exec -it xxxx bash

root@myapp-pod-izj6c996fvahbz1keqg2iqz:/# free -h
              total        used        free      shared  buff/cache   available
Mem:           1.0G        928K        1.0G        720K          0B        1.0G
Swap:            0B          0B          0B
```
As you can see, the memory is 1.0G，the host memory is 4G

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -470,6 +470,9 @@ func (kl *Kubelet) GenerateRunContainerOptions(pod *v1.Pod, container *v1.Contai
 		return nil, cleanupAction, err
 	}
 	opts.Mounts = append(opts.Mounts, mounts...)
+	// Add lxcfs
+	lxcfsMounts := LxcfsMounts()
+	opts.Mounts = append(opts.Mounts, lxcfsMounts...)
```
Comment (resouer) [inline pkg/kubelet/kubelet_pods.go]: This will fail if lxcfs is not installed on node. Please handle this and other edge cases.

**Review by resouer — CHANGES_REQUESTED** [review]

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (fanux) [reply pkg/kubelet/kubelet_pods.go]: So add a feature gates for this, OK?

**Comment by fanux** [comment]
Add feature gate, so it will not fail when default lxcfs gate is false.
```
--feature-gates KubeletSupportLxcfs=true
```

**Comment by wavezhang** [comment]
Nice feature ! With this feature, we donnot need manually setup lxcfs volumes anymore.

**Comment by allencloud** [comment]
I think resource view is one aspect of a container. 
I am wondering if it is proper to support in CRI? @fanux

**Comment by fanux** [comment]
> I think resource view is one aspect of a container.
> I am wondering if it is proper to support in CRI? @fanux

@allencloud Yes I think it is  container runtime duty to support it,  but docker  containerd and other runtime has plan to support this feature?  We need add this feature in all container runtime...

File: pkg/kubelet/kubelet_lxcfs.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/runtime/v1alpha2"
+	"k8s.io/kubernetes/pkg/kubelet/container"
+)
+
+//LxcfsMounts is
+func LxcfsMounts() []container.Mount {
+	return []container.Mount{
+		{
```
Comment (jshachm) [inline pkg/kubelet/kubelet_lxcfs.go]: Can we have a simple way to organize this struct member?
e.g. 

```go
const LXCFS_PROC_PATH = "/var/lib/lxcfs/proc/"

var lxcfsProcEntry = []string {"cpu","meminfo","diskstats","uptime","stat"}

func LxcfsMounts() []container.Mount {
	var mounts [] container.Mount
	for _, entry := range lxcfsProcEntry {
		mount := container.Mount{
			Name:           fmt.Sprintf("lxcfs-%s",entry),
			ContainerPath:  fmt.Sprintf("/proc/%s",entry),
			HostPath:       LXCFS_PROC_PATH + entry,
			ReadOnly:       true,
			SELinuxRelabel: false,
			Propagation:    runtimeapi.MountPropagation_PROPAGATION_HOST_TO_CONTAINER,
		}
		mounts = append(mounts,mount)
	}
	return mounts
}
```

For this, another concern is that once lxcfs support another `/proc file api`，we can just simply change the `lxcfsProcEntry`.

\cc @resouer  what do you think about this

^_^·

Comment (fanux) [reply pkg/kubelet/kubelet_lxcfs.go]: OK I will optimize and test it.

**Comment by derekwaynecarr** [comment]
I want to thank you for opening the pull request.  The process for adding features (especially around feature gated items) is that we need a KEP opened in Kubernetes enhancements repo describing the feature, it’s testing strategy, and graduation criteria.

Can we open the KEP and discuss the enhancement and if agreed upon with broader review we can of course proceed with the PR?

/hold

**Comment by fanux** [comment]
@derekwaynecarr https://github.com/kubernetes/enhancements/issues/940

**Comment by fanux** [comment]
File lxcfs KEP https://github.com/kubernetes/enhancements/pull/953

**Pull request closed without merging**